include/openthread/heap.h

```diff
@@ -0,0 +1,99 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the API for setting external heap in OpenThread.
+ */
+
+#ifndef OPENTHREAD_HEAP_H_
+#define OPENTHREAD_HEAP_H_
+
+#include <openthread/instance.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-heap
+ *
+ * @brief
+ *   This module includes functions that set the external OpenThread heap.
+ *
+ * @{
+ *
+ */
+
+/**
+ * Function pointer used to set external CAlloc function for OpenThread.
+ *
+ * @param[in]   aCount  Number of allocate units.
+ * @param[in]   aSize   Unit size in bytes.
+ *
+ * @returns A pointer to the allocated memory.
+ *
+ * @retval  NULL    Indicates not enough memory.
+ *
+ */
+typedef void *(*otHeapCAllocFn)(size_t aCount, size_t aSize);
```

Comment:
For extensibility, how about we define a structure to accept parameters for initializing a OpenThread instance? Then these two pointers can be delivered before calling Instance constructor?

That is, we'll extend:
```diff
// include/openthread/instance.h


- otInstance *otInstanceInitSingle(void);

+ typedef struct otConfig {
+    otHeapCAllocFn aCAlloc;
+    otHeapFreeFn aFree;
+ } otConfig;
+ otInstance *otInstanceInitSingle(const otConfig *aConfig);
```
Then in the Instance class, we define two inline functions: `Instance::HeapCAlloc(), Instance::HeapFree()`, which calls mHeap.CAlloc() and mHeap.Free() if `OPENTHREAD_CONFIG_EXTERNAL_HEAP_ENABLE` is not defined, otherwise, two member variables (mFree, mCAlloc) are defined, and `Instance::HeapCAlloc(), Instance::HeapFree()` calls the member function pointers.

I prefer this way because:

1. No need to introduce a new class ExternalHeap and many dummy functions.
2. No need to move code of `Heap`
3. We can add more fields into `otConfig` for future requirements.

Since this involves the API change, what do you think @jwhui @abtink @LuDuda

Reply:
I think that this type of configuration suits the platform layer better. But moving heap to the platform would require more effort and we should plan it for the future. Also, we may try to remove the heap requirement from the border agent - we can reuse buffer allocation used by Message. This way we are left with the heap requirement only in mbedTLS - and this can be done during system init or even before it.

Reply:
I agree with @KaSroka that heap management should be rather specific to **system layer** (e.g. FreeRTOS, Zephyr, Bare-Metal etc.). So long run, maybe we should think of the extraction of the next layer (next to platform layer). The downside is that currently, OpenThread core defines the size of the heap taking into account enabled features (e.g. CoAPs, DTLS..).

Personally, I would vote to keep the current (this PR) approach since it does not touch an existing API.

Later on, maybe rewrite current InternalHeap implementation to C language and put it into `examples/utils` folder (same as it is done with settings implementation). Then the platform/system layer would be responsible for delivering Heap API. 

@abtink @jwhui this PR is needed to enable thread-safe operations in RTOS environment e.g. FreeRTOS (OpenWeave) can you help to move it forward?

Reply:
This PR is for enabling two functions to manage heap memory, I am wondering if we have to interface at the `Heap` class level, which results in most methods dummy.

I feel it hard for me to write in English, but maybe code is more readable for us. So I pulled this PR and made some adjustments.
https://github.com/openthread/openthread/compare/master...bukepo:utils/external-heap
Could you check the link and see if I that works for you? @LuDuda @KaSroka @e-rk

Reply:
I like this approach. It seems much cleaner and is functionally identical to what I implemented.

Reply:
Hello @bukepo, sorry for misunderstanding I agree with the changes you proposed, I think @e-rk planned and should include them in this PR.

My and Kamil's comments were only towards adding the `otConfig` structure - which I believe is a good solution and we can go with it, but just wondering if it fits in the overall architecture.

Reply:
Great! And thank you for the understanding!